src/hdmf/common/table.py

```diff
@@ -1421,6 +1421,26 @@ def __repr__(self):
                                                               id(self.table))
         return template
 
+    def _validate_on_set_parent(self):
+        # when this DynamicTableRegion is added to a parent, check:
+        # 1) if the table was read from a written file, no need to validate further
+        p = self.table
+        while p is not None:
+            if p.container_source is not None:
```

Comment:
@rly what's the difference between parent and container source? Is the source meant to be the root?

Reply:
This checks if the table being used in the DTR has a source, if it does then pass, if not then it will check the parent for a source (and so on)

Reply:
Container_source ==> path to the file ==>Does zarrio set container_source (confirm?)

Reply:
container_source is set in `BuildManager.build` . ZarrIO deals with read/write of Builders, but the logic for creating Containters is the same as in HDMF, i.e., HDMF is setting the container_source and the logic to do so is independent of whether HDF5IO or ZarrIO is being used.